Fix NativeFunctions.h for c10-full ops

smessmer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #46092 Make empty c10-full
* #46091 Fix device guard for c10-full ops
* **#46090 Fix NativeFunctions.h for c10-full ops**

Differential Revision: [D24219942](https://our.internmc.facebook.com/intern/diff/D24219942/)

smessmer committed:
Fix NativeFunctions.h for c10-full ops



Differential Revision: [D24219942](https://our.internmc.facebook.com/intern/diff/D24219942/)

[ghstack-poisoned]

ezyang commented on tools/codegen/api/dispatcher.py:
```diff
@@ -61,10 +61,11 @@ def argument(a: Argument) -> DispatcherArgument:
     else:
         assert local.use_c10_dispatcher() is UseC10Dispatcher.with_codegenerated_unboxing_wrapper
         la = legacy_dispatcher.argument(a)
+        assert len(la) == 1, "TensorOptions arguments not supported"
```
IIUC, the real invariant here is that when `local.use_c10_dispatcher() is UseC10Dispatcher.with_codegenerated_unboxing_wrapper` then `legacy_dispatcher.argument(a)` is guaranteed to return a single element; it has nothing to do with TensorOptions arguments not being supported.

Although this is ok to leave up to personal preference, I think a better way to handle the problem here is to split `legacy_dispatcher.argument` into two functions: one that is guaranteed to return only a single LegacyDispatcherArgument (but is only valid for `with_codegenerated_unboxing_wrapper`), and then another one which is valid in all cases but might return a sequence.

ezyang commented on tools/codegen/api/legacy_dispatcher.py:
```diff
@@ -38,39 +39,68 @@ def returns_type(rs: Sequence[Return]) -> str:
+                    name='pin_memory',
+                    default='{}',
+                    argument=a,
+                )]
```
And now we pray that this doesn't need to be packed (in the same way CppArguments need it)......

smessmer committed:
Update on "Fix NativeFunctions.h for c10-full ops"


Differential Revision: [D24219942](https://our.internmc.facebook.com/intern/diff/D24219942/)

[ghstack-poisoned]

smessmer committed:
Update on "Fix NativeFunctions.h for c10-full ops"


Differential Revision: [D24219942](https://our.internmc.facebook.com/intern/diff/D24219942/)

[ghstack-poisoned]

smessmer committed:
Update on "Fix NativeFunctions.h for c10-full ops"


Differential Revision: [D24219942](https://our.internmc.facebook.com/intern/diff/D24219942/)

[ghstack-poisoned]

smessmer committed:
Update on "Fix NativeFunctions.h for c10-full ops"


Differential Revision: [D24219942](https://our.internmc.facebook.com/intern/diff/D24219942/)

[ghstack-poisoned]

gchanan commented:
"Fix" -- what's broken?
